pkg/schedule/affinity/policy_test.go

```diff
@@ -0,0 +1,192 @@
+
+	// First observation makes group effective with store 1.
+	region1 := core.NewRegionInfo(
+		&metapb.Region{
+			Id:       10,
+			StartKey: []byte(""),
+			EndKey:   []byte("a"),
+			Peers:    []*metapb.Peer{{Id: 11, StoreId: 1, Role: metapb.PeerRole_Voter}},
+		},
+		&metapb.Peer{Id: 11, StoreId: 1, Role: metapb.PeerRole_Voter},
+	)
+	manager.ObserveAvailableRegion(region1, manager.GetAffinityGroupState("g"))
+	state := manager.GetAffinityGroupState("g")
+	re.True(state.Effect)
+	re.Equal(uint64(1), state.LeaderStoreID)
+	re.ElementsMatch([]uint64{1}, state.VoterStoreIDs)
+
+	// Second observation with different layout should not overwrite.
+	region2 := core.NewRegionInfo(
+		&metapb.Region{
+			Id:       20,
+			StartKey: []byte("a"),
+			EndKey:   []byte("b"),
+			Peers:    []*metapb.Peer{{Id: 21, StoreId: 2, Role: metapb.PeerRole_Voter}},
+		},
+		&metapb.Peer{Id: 21, StoreId: 2, Role: metapb.PeerRole_Voter},
+	)
+	manager.ObserveAvailableRegion(region2, manager.GetAffinityGroupState("g"))
+	state2 := manager.GetAffinityGroupState("g")
+	re.True(state2.Effect)
+	re.Equal(uint64(1), state2.LeaderStoreID)
+	re.ElementsMatch([]uint64{1}, state2.VoterStoreIDs)
+}
+
+func TestAvailabilityCheckInvalidatesGroup(t *testing.T) {
+	re := require.New(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	store := storage.NewStorageWithMemoryBackend()
+	storeInfos := core.NewStoresInfo()
+	store1 := core.NewStoreInfo(&metapb.Store{Id: 1, Address: "s1"})
+	store1 = store1.Clone(core.SetLastHeartbeatTS(time.Now()))
+	storeInfos.PutStore(store1)
+	store2 := core.NewStoreInfo(&metapb.Store{Id: 2, Address: "s2"})
+	store2 = store2.Clone(core.SetLastHeartbeatTS(time.Now()), core.SetNodeState(metapb.NodeState_Removing))
+	storeInfos.PutStore(store2)
+
+	conf := mockconfig.NewTestOptions()
+	manager := NewManager(ctx, store, storeInfos, conf, nil)
+	re.NoError(manager.Initialize())
+
+	group := &Group{ID: "avail", LeaderStoreID: 1, VoterStoreIDs: []uint64{1, 2}}
+	re.NoError(manager.SaveAffinityGroups([]GroupWithRanges{{Group: group}}))
+	_, err := manager.UpdateGroupPeers("avail", 1, []uint64{1, 2})
+	re.NoError(err)
+	state := manager.GetAffinityGroupState("avail")
+	re.True(state.Effect)
+
+	// Simulate store 2 unavailable.
+	unavailable := map[uint64]storeState{2: removingOrRemoved}
+	manager.setUnavailableStores(unavailable)
+
+	state2 := manager.GetAffinityGroupState("avail")
+	re.False(state2.Effect)
+}
+
+func TestStoreHealthCheck(t *testing.T) {
+	re := require.New(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Create a test storage
+	store := storage.NewStorageWithMemoryBackend()
+
+	// Create mock stores
+	storeInfos := core.NewStoresInfo()
+	store1 := core.NewStoreInfo(&metapb.Store{Id: 1, Address: "test1"})
+	store2 := core.NewStoreInfo(&metapb.Store{Id: 2, Address: "test2"})
+	store3 := core.NewStoreInfo(&metapb.Store{Id: 3, Address: "test3"})
```

Comment:
```suggestion
	store1 := core.NewStoreInfo(&metapb.Store{Id: 1, Address: "test1", NodeState: metapb.NodeState_Serving})
	store2 := core.NewStoreInfo(&metapb.Store{Id: 2, Address: "test2", NodeState: metapb.NodeState_Serving})
	store3 := core.NewStoreInfo(&metapb.Store{Id: 3, Address: "test3", NodeState: metapb.NodeState_Serving})
```
Otherwise, the test will fail